spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -226,6 +222,12 @@ public Scan build() {
       scan = scan.asOfTime(asOfTimestamp);
     }
 
+    if (branch != null) {
+      scan.useRef(branch);
+    } else if (tag != null) {
+      scan.useRef(tag);
+    }
```

Comment:
I think the issue is that scan.useRef(tag) by itself doesn't update the snapshot in the context; it's a builder like pattern so it needs to be `scan = scan.useRef(ref)`. That's why the tests are failing; the context snapshot isn't set, so by default the scan reads main

Reply:
my bad missed this! Thanks for pointing out!

Reply:
No problem!